### PR TITLE
Upgrade core version to latest release 3.1.401

### PR DESCRIPTION
**Pull request by JialinXin**
### Summary of the changes (Less than 80 chars)
Upgrade build dotnet core sdk version to latest release 3.1.401.

### CODE REVIEW

File: test/Microsoft.Azure.SignalR.Common.Tests/BackOffPolicyFacts.cs [hunk 1]
```diff
@@ -180,8 +180,9 @@ static async Task RunProbeTests(TestData testData)
                     {
                         // This effectively defines the order of the calls
                         await Task.Delay(currentProbe.InitialDelay);
-
-                        testData.Results[index].ActualResult = await policy.CallProbeWithBackOffAsync(probeFunc, testData.BkOffFunc);
```
Comment (JialinXin) [inline test/Microsoft.Azure.SignalR.Common.Tests/BackOffPolicyFacts.cs:184]: @KKhurin No idea why it's not working in original way under 3.1.401 that `ObjectNullReference` throws. Would you help take a look? Now workaround by separate the steps. Thanks to @vicancy!

Comment (KKhurin) [reply test/Microsoft.Azure.SignalR.Common.Tests/BackOffPolicyFacts.cs:184]: To my not so sharp eyes this looks like a jitter bug creeping into the new compiler code :( 

When reproducing in windbg I see that the compiler incorrectly captures testData.Results[index] in `BackOffPolicyFacts+<>c__DisplayClass18_1+<<RunProbeTests>b__1>d`  **_before_** awaiting for the task returned from `policy.CallProbeWithBackOffAsync`. 

Here is how it looks like:
~~~
0:020> .lastevent
Last event: 290c.1bc8: Access violation - code c0000005 (first chance)

0:020> r
rax=0000000000000001 rbx=000001ff33a5f518 rcx=0000000000000000
rdx=000001ff33a5f518 rsi=000001ff33a5ea90 rdi=000000c987a7ded0
rsp=000000c987a7de30 rbp=000000c987a7dee0
rip=00007ffa979b92e4

0:020> !u 00007ffa979b92e4
~~~
skipping miles of code until we await for the task returned at line 184
~~~
00007ffa`979b92c2 e8291678fe      call    CLRStub[MethodDescPrestub]@7ffa9613a8f0 (00007ffa`9613a8f0) (System.Runtime.CompilerServices.TaskAwaiter`1[[System.Boolean, System.Private.CoreLib]].GetResult(), mdToken: 0000000006003730)
00007ffa`979b92c7 8945a4          mov     dword ptr [rbp-5Ch],eax
00007ffa`979b92ca 488b4da8        mov     rcx,qword ptr [rbp-58h]
00007ffa`979b92ce 8b45a4          mov     eax,dword ptr [rbp-5Ch]
00007ffa`979b92d1 884124          mov     byte ptr [rcx+24h],al
00007ffa`979b92d4 488b4d10        mov     rcx,qword ptr [rbp+10h]
00007ffa`979b92d8 488b4910        mov     rcx,qword ptr [rcx+10h]
00007ffa`979b92dc 488b4510        mov     rax,qword ptr [rbp+10h]
00007ffa`979b92e0 0fb64024        movzx   eax,byte ptr [rax+24h]
>>> 00007ffa`979b92e4 884118          mov     byte ptr [rcx+18h],al
~~~
So the problem is `rcx` is 0. Following where it came from and dumping ` [rbp+10h]` 
~~~
0:020> dd 000000c987a7dee0+10h
000000c9`87a7def0  33a5ea90 000001ff 

0:020> !do 000001ff33a5ea90 
Name:        Microsoft.Azure.SignalR.Common.Tests.BackOffPolicyFacts+<>c__DisplayClass18_1+<<RunProbeTests>b__1>d
MethodTable: 00007ffa962bd338
EEClass:     00007ffa962c9148
Size:        72(0x48) bytes
File:        C:\Users\kkhurin\source\repos\AllSDK\31\azure-signalr\test\Microsoft.Azure.SignalR.Common.Tests\bin\Debug\netcoreapp3.1\Microsoft.Azure.SignalR.Common.Tests.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007ffa95bab1e8  4000081       20         System.Int32  1 instance               -1 <>1__state
00007ffa962baa10  4000082       28 ...TaskMethodBuilder  1 instance 000001ff33a5eab8 <>t__builder
00007ffa962bc010  4000083        8 ..._DisplayClass18_1  0 instance 000001ff33a5e9e0 <>4__this
00007ffa962bb318  4000084       10 ...Facts+ProbeResult  0 instance 0000000000000000 <>s__1
00007ffa95ba71c8  4000085       24       System.Boolean  1 instance                1 <>s__2
00007ffa95c64170  4000086       18     System.Exception  0 instance 0000000000000000 <ex>5__3
00007ffa962bb840  4000087       30 ...vices.TaskAwaiter  1 instance 000001ff33a5eac0 <>u__1
00007ffa962bd1f8  4000088       38 ...Private.CoreLib]]  1 instance 000001ff33a5eac8 <>u__2
~~~

So the locally captured `ProbeResult` from `testData.Results[index]` before the await is null 
(offset `10 ...Facts+ProbeResult  0 instance 0000000000000000 <>s__1`)
The actual 'testData.Results[0]' is not null:
~~~
 !DumpObj /d 000001ff33a5e9e0
Name:        Microsoft.Azure.SignalR.Common.Tests.BackOffPolicyFacts+<>c__DisplayClass18_1
MethodTable: 00007ffa962bc010
EEClass:     00007ffa962c8a68
Size:        48(0x30) bytes
File:        C:\Users\kkhurin\source\repos\AllSDK\31\azure-signalr\test\Microsoft.Azure.SignalR.Common.Tests\bin\Debug\netcoreapp3.1\Microsoft.Azure.SignalR.Common.Tests.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007ffa95bab1e8  400005d       20         System.Int32  1 instance                0 index
00007ffa962bb208  400005e        8 ...yFacts+ProbeParam  0 instance 000001ff33a5e7c0 currentProbe
00007ffa97e91f70  400005f       10 ...Private.CoreLib]]  0 instance 000001ff33a5ea10 probeFunc
00007ffa962bbee0  4000060       18 ..._DisplayClass18_0  0 instance 000001ff33a5e930 CS$<>8__locals1

0:020> !DumpObj /d 000001ff33a5e930
Name:        Microsoft.Azure.SignalR.Common.Tests.BackOffPolicyFacts+<>c__DisplayClass18_0
MethodTable: 00007ffa962bbee0
EEClass:     00007ffa962c89f0
Size:        48(0x30) bytes
File:        C:\Users\kkhurin\source\repos\AllSDK\31\azure-signalr\test\Microsoft.Azure.SignalR.Common.Tests\bin\Debug\netcoreapp3.1\Microsoft.Azure.SignalR.Common.Tests.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007ffa962bb418  4000059        8 ...icyFacts+TestData  0 instance 000001ff33a5e760 testData
00007ffa95ce6720  400005a       20      System.DateTime  1 instance 000001ff33a5e950 startTime
00007ffa95bab1e8  400005b       18         System.Int32  1 instance                1 _funcCallNumber
00007ffa969963e0  400005c       10 ...mon.BackOffPolicy  0 instance 000001ff33a5e960 policy

0:020> !DumpObj /d 000001ff33a5e760
Name:        Microsoft.Azure.SignalR.Common.Tests.BackOffPolicyFacts+TestData
MethodTable: 00007ffa962bb418
EEClass:     00007ffa962c8618
Size:        48(0x30) bytes
File:        C:\Users\kkhurin\source\repos\AllSDK\31\azure-signalr\test\Microsoft.Azure.SignalR.Common.Tests\bin\Debug\netcoreapp3.1\Microsoft.Azure.SignalR.Common.Tests.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007ffa97e5eaf0  4000040        8 ...acts+ProbeParam[]  0 instance 000001ff33a5e790 Params
00007ffa97e5ef08  4000041       10 ...Private.CoreLib]]  0 instance 000001ff33a5e880 BkOffFunc
00007ffa97e5ecd0  4000042       18    System.TimeSpan[]  0 instance 000001ff33a5e838 ExpectedCallTimes
00007ffa97e91bd8  4000043       20 ...cts+ProbeResult[]  0 instance 000001ff33a5e980 Results

0:020> !DumpArray /d 000001ff33a5e980
Name:        Microsoft.Azure.SignalR.Common.Tests.BackOffPolicyFacts+ProbeResult[]
MethodTable: 00007ffa97e91bd8
EEClass:     00007ffa95ba6578
Size:        48(0x30) bytes
Array:       Rank 1, Number of elements 3, Type CLASS
Element Methodtable: 00007ffa962bb318
[0] 000001ff33a5ebf0
[1] null
[2] null

!DumpObj /d 000001ff33a5ebf0
Name:        Microsoft.Azure.SignalR.Common.Tests.BackOffPolicyFacts+ProbeResult
MethodTable: 00007ffa962bb318
EEClass:     00007ffa962c85a0
Size:        48(0x30) bytes
File:        C:\Users\kkhurin\source\repos\AllSDK\31\azure-signalr\test\Microsoft.Azure.SignalR.Common.Tests\bin\Debug\netcoreapp3.1\Microsoft.Azure.SignalR.Common.Tests.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007ffa95ba71c8  400003b       18       System.Boolean  1 instance                0 ActualResult
00007ffa95c64170  400003c        8     System.Exception  0 instance 0000000000000000 ActualException
00007ffa95bab1e8  400003d       10         System.Int32  1 instance                1 ActualCallOrder
00007ffa95bab1e8  400003e       14         System.Int32  1 instance                1 ActualNumberOfCalls
00007ffa95e479c0  400003f       20      System.TimeSpan  1 instance 000001ff33a5ec10 ActualCallTime
~~~

So the bug in jitter is that it didn't take into account the possibility of us setting 'testData.Results[0]' inside `probeFunc`. 
It captured the address of 'testData.Results[0]' (which was 0) before the await and ran into a null ref.

A more robust workaround would be to move `testData.Results[index] = new ProbeResult()` from line 166 to inside the `testFunc` so we won't rely on the jitter to correctly capturing and addressing the variables inside async methods.

File: test/Microsoft.Azure.SignalR.Common.Tests/BackOffPolicyFacts.cs [hunk 2]
```diff
@@ -181,7 +181,7 @@ static async Task RunProbeTests(TestData testData)
                     {
                         // This delay effectively defines the order of the calls
                         await Task.Delay(currentProbe.InitialDelay);
-                        testData.Results[index].ActualResult = policy.CallProbeWithBackOffAsync(probeFunc, testData.BkOffFunc);
+                        testData.Results[index].ActualResult = await policy.CallProbeWithBackOffAsync(probeFunc, testData.BkOffFunc);
```
Comment (KKhurin) [inline test/Microsoft.Azure.SignalR.Common.Tests/BackOffPolicyFacts.cs:184]: 👍

**Review by vicancy — APPROVED** [review]

**Comment by jaredpar** [comment]
The current behavior, ` testData.Results[index]`, being evaluated before the `await` is the expected behavior here.

**Comment by KKhurin** [comment]
Thanks, @jaredpar 

So this makes 2 snippets of code like the following not equivalent?
~~~
var r = await AsyncMethod();
testData.Results[index] = r;
~~~
vs
~~~
testData.Results[index] = await AsyncMethod();
~~~

**Comment by jaredpar** [comment]
Correct. Have a gist that demonstrates this behavior

https://gist.github.com/jaredpar/8c485c80b843745eabbf23d86e24c791

**Comment by vicancy** [comment]
Shall we file a bug for this behavior @jaredpar @KKhurin ?

**Comment by jaredpar** [comment]
@vicancy This behavior is expected so there is no need for a bug.

**Pull request merged**